### DISCUSSION TITLE
SpringBoot项目log4j2.xml配置放到apollo是否支持？

### DISCUSSION THREAD

**Discussion by hongkuan45** [apolloconfig/apollo#4245] [Q&A]
期望的用法是，在apollo创建一个namaspace （比如命名为log4j2.xml)。 
启动项目指定apollo的log4j2.xml为日志配置文件。
以此做到在线化log配置。

此文做了自定义开发实现了相同的目的。想了解有没有更原生的方式。
https://blog.csdn.net/xiaokuanblog/article/details/123114628

**Comment by nisiyong** [comment]
We had supported this feature in our company.
I do this by log4j2 origin way, extend log4j2 `org.apache.logging.log4j.core.config.ConfigurationFactory` in the dependency module `apollo-log4j2`.

## Usage

**1) maven dependency**

if someone want to manage log4j2.xml in apollo, they should add this dependency:
```xml
<dependency>
    <groupId>com.ctrip.framework.apollo</groupId>
    <artifactId>apollo-log4j2</artifactId>
</dependency>
```

**2) create log4j2.xml namespace**
<img width="681" alt="image" src="https://user-images.githubusercontent.com/8198862/156280242-dcfd2cd7-f211-47f8-bdd1-f2a343a897bd.png">

**3) enable this feature by system properties**

```
-Dapollo.log4j2.enabled=true
```
> Note: By default, it will load the local log4j2.xml in the classpath.

---

If anyone wants this feature, please comment on this issue. I think I could create a PR for this, we have already used this feature at least a year.
<img width="369" alt="image" src="https://user-images.githubusercontent.com/8198862/156281260-9c08168e-d153-4bde-8d9c-eafa633c0924.png">

@nobodyiam I create an `apollo-plugin` module and put it in this.

**Reply by nobodyiam** [reply]
@nisiyong This looks cool, maybe we could create a standalone repository to host apollo plugins?

**Reply by nisiyong** [reply]
@nobodyiam It is inspired from [dubbo-plugin](https://github.com/apache/dubbo/tree/3.0/dubbo-plugin).

It is fine to me whether to separate it to an individual repository, I think the only problem is whether it should keep the same maven version with apollo-client.

**Reply by nobodyiam** [reply]
I think it doesn't need to keep the same maven version with apollo-client as long as the apollo-client keeps backward compatibility.

**Reply by nisiyong** [reply]
Okay, I think we could kick off a new project to do this, and named it like `apollo-plugins`, `apollo-extensions`, or `apollo-externals`.

**Reply by nobodyiam** [reply]
That sounds like a great idea!

**Reply by nisiyong** [reply]
Which name do you prefer? I can create it in the next few days.

**Reply by Anilople** [reply]
How about refer to maven's plugins? https://mvnrepository.com/artifact/org.apache.maven.plugins

All maven plugin's parent is 

```xml
  <parent>
    <groupId>org.apache.maven.plugins</groupId>
    <artifactId>maven-plugins</artifactId>
    <version>34</version>
    <relativePath/>
  </parent>
```
and their artifactId is maven-xxx-plugin, for example
```xml
<artifactId>maven-dependency-plugin</artifactId>
```
---

for the log4j2, can we use name convention like follow?
```xml
<dependency>
    <groupId>com.ctrip.framework.apollo</groupId>
    <artifactId>apollo-log4j2-plugin</artifactId>
</dependency>
```
and it's parent is
```xml
  <parent>
    <groupId>com.ctrip.framework.apollo</groupId>
    <artifactId>apollo-plugins</artifactId>
  </parent>
```
or other better name convention?

**Reply by nisiyong** [reply]
In my opinion, if we create a new repo to do this, it is not only plugin things.
We could put other third parties feature stuff in this repo, it is more like apollo-externals, inspired by this project [rocketmq-externals](https://github.com/apache/rocketmq-externals).
